File: app/src/main/java/org/bitcoindotcom/bchprocessor/bip70/Bip70Manager.java
package org.bitcoindotcom.bchprocessor.bip70;

import android.app.Application;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

public class Bip70Manager {
    public final Application app;
    public WebSocketHandler socketHandler;

    public Bip70Manager(Application app) {
        this.app = app;
    }

    public void reconnectIfNecessary() {
        final ConnectivityManager connectivityManager = (ConnectivityManager) app.getSystemService(Context.CONNECTIVITY_SERVICE);
        if (connectivityManager == null) {
            return;
        }
        final NetworkInfo ni = connectivityManager.getActiveNetworkInfo();
        if (ni != null && ni.isConnectedOrConnecting()) {
            if (socketHandler != null && !socketHandler.isConnected()) {
                socketHandler.start();
            }
        }
        if (socketHandler != null) {
            WebSocketHandler.notifyConnectionStatus(app, socketHandler.isConnected());
        }
    }

    public void startWebsockets(String invoiceId) {
        socketHandler = new Bip70SocketHandler(app, invoiceId);
        socketHandler.start();
    }

    public void stopSocket() {
        if (socketHandler != null) {
            socketHandler.stop();
            socketHandler = null;
        }
    }
}

File: app/src/main/java/org/bitcoindotcom/bchprocessor/bip70/Bip70Manager.kt
package org.bitcoindotcom.bchprocessor.bip70

import android.app.Application
import android.content.Context
import android.net.ConnectivityManager
import android.util.Log
import java.lang.Boolean.FALSE

class Bip70Manager(val app: Application) {
    var socketHandler: Bip70SocketHandler? = null
    fun reconnectIfNecessary() {
        val connectivityManager = app.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
        val networkInfo = connectivityManager?.activeNetworkInfo
        val ws = socketHandler
        if (networkInfo?.isConnectedOrConnecting ?: FALSE) {
            if (ws != null && !ws.isConnected) {
                ws.start()
            }
        }
        if (ws != null) {
            WebSocketHandler.notifyConnectionStatus(app, ws.isConnected)
        }
    }

    fun startWebsockets(invoiceId: String) {
        val ms = System.currentTimeMillis()
        socketHandler = Bip70SocketHandler(app, invoiceId)
        socketHandler?.start()
        Log.d(WebSocketHandler.TAG, "startWebsockets for invoice:$invoiceId in ${System.currentTimeMillis()-ms} ms")
    }

    fun stopSocket() {
        val ms = System.currentTimeMillis()
        val invoiceId = socketHandler?.invoiceId
        socketHandler?.stop()
        socketHandler = null
        Log.d(WebSocketHandler.TAG, "stopWebsockets for invoice:$invoiceId in ${System.currentTimeMillis()-ms} ms")
    }
}